Interpolated string handlers in the CFG

333fred opened the pull request:
Closes https://github.com/dotnet/roslyn/issues/54718. Implements support for interpolated string handlers in the CFG. For ease of review, I recommend reviewing commits 1 and 2 separately: commit 1 implements lowering for interpolated string handlers without supporting containing context. Commit 2 adds the machinery to set up the containing context and uses it everywhere we have arguments in the IOperation machinery.

333fred commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:
```diff
@@ -2268,19 +2268,1788 @@ public void InterpolatedStringHandlerConversionFlow_01()
+}
+";
+
+            /* TODO after merge
```
This diagnostic will be reported after https://github.com/dotnet/roslyn/pull/57456 is merged.

333fred commented:
@AlekseyTs @chsienki for review of this PR.

333fred commented:
https://github.com/dotnet/roslyn/issues/57484 covers the API proposal for `IsInitialized`.

chsienki commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6344,167 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                }
+
+                var stack = ArrayBuilder<IInterpolatedStringAdditionOperation>.GetInstance();
+                pushLeftNodes((IInterpolatedStringAdditionOperation)creation.Content, stack);
```
Consider asserting that this is an `IInterpolatedStringAdditionOperation` beforehand

333fred replied:
I don't think that's particularly necessary. The code will just fail here.

chsienki commented on src/Compilers/Core/Portable/Operations/OperationNodes.cs:
```diff
@@ -550,4 +550,8 @@ private BlockOperation(ImmutableArray<IOperation> statements, SemanticModel sema
         }
     }
+
+    internal sealed partial class FlowCaptureReferenceOperation
```
Why is this needed?

chsienki commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2044,9 +2044,161 @@ private IOperation PeekOperand()
+                                switch (argumentPlaceholder.PlaceholderKind)
+                                {
+                                    case InterpolatedStringArgumentPlaceholderKind.CallsiteReceiver:
+                                        if (capturedInstance != null)
```
`if (capturedInstance is not null)`

chsienki commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2044,9 +2044,161 @@ private IOperation PeekOperand()
+
+                placeholderKeys.Free();
+
+                if (capturedInstance != null)
```
`is not null`

333fred commented:
@AlekseyTs @chsienki please take another look.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2045,232 @@ private IOperation PeekOperand()
+                {
+                    return args.Any(
+                        static (arg, unwrapper) => unwrapper(arg) switch
+                {
```
>{ [](http://example.com/codeflow?start=16&length=1)

The alignment looks wrong.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2045,232 @@ private IOperation PeekOperand()
+                    IInterpolatedStringHandlerArgumentPlaceholderOperation { PlaceholderKind: not InterpolatedStringArgumentPlaceholderKind.TrailingValidityArgument } => true,
+                    IConversionOperation
+                    {
+                        Operand: IInterpolatedStringHandlerArgumentPlaceholderOperation { PlaceholderKind: not InterpolatedStringArgumentPlaceholderKind.TrailingValidityArgument }
```
>IInterpolatedStringHandlerArgumentPlaceholderOperation [](http://example.com/codeflow?start=33&length=54)

This looks fragile. I think when user-defined conversions are involved, original operand might not be the immediate operand of a top level conversion.

AlekseyTs replied:
>not InterpolatedStringArgumentPlaceholderKind.TrailingValidityArgument [](http://example.com/codeflow?start=107&length=70)

Can this check ever fail?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2045,232 @@ private IOperation PeekOperand()
+                return arguments.Any(
+                    static arg => arg.Value switch
+                    {
+                        IInterpolatedStringHandlerCreationOperation { HandlerCreation: IObjectCreationOperation { Arguments: { } creationArgs } } =>
```
>IInterpolatedStringHandlerCreationOperation [](http://example.com/codeflow?start=24&length=43)

I do not see any benefit from repeating this type check for every case within the switch. It feels like the code will be easier to follow when the type check is moved outside and performed once.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2045,232 @@ private IOperation PeekOperand()
+                    PushOperand(VisitRequired(UnwrapArgumentDoNotCaptureDirectly(arguments[i])));
+                }
+
+                bool visitArgument<T>(IOperation originalArgument, ImmutableArray<T> nestedArguments, Func<T, IOperation> unwrapper) where T : IOperation
```
>IOperation [](http://example.com/codeflow?start=38&length=10)

IInterpolatedStringHandlerCreationOperation?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -75,6 +77,25 @@ private RegionBuilder CurrentRegionRequired
+            return _placeholderDictionary[key];
+        }
+
+        private void AddPlaceholder(IOperation key, IOperation value)
```
>AddPlaceholder [](http://example.com/codeflow?start=21&length=14)

AddPlaceholderReplacement?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -75,6 +77,25 @@ private RegionBuilder CurrentRegionRequired
             }
         }
 
+        private IOperation GetPlaceholder(IOperation key)
```
>GetPlaceholder [](http://example.com/codeflow?start=27&length=14)

GetPlaceholderReplacement?

333fred replied:
These are no longer present.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -75,6 +77,25 @@ private RegionBuilder CurrentRegionRequired
+            (_placeholderDictionary ??= PooledDictionary<IOperation, IOperation>.GetInstance()).Add(key, value);
+        }
+
+        private void RemovePlaceholder(IOperation key)
```
>RemovePlaceholder [](http://example.com/codeflow?start=21&length=17)

RemovePlaceholderReplacement?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2045,232 @@ private IOperation PeekOperand()
+
+                    IOperation visitedArgument = VisitRequired(originalArgument);
+                    Debug.Assert(visitedArgument is IFlowCaptureReferenceOperation);
+                    visitedArguments.Add(visitedArgument);
```
>visitedArguments.Add(visitedArgument); [](http://example.com/codeflow?start=20&length=38)

I think the logic would be simpler to follow if we were to limit the purpose of this method to visiting IInterpolatedStringHandlerCreationOperation argument and returning it. The caller can handle adition of the result to the list.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2045,232 @@ private IOperation PeekOperand()
+
+                    placeholderKeys.Clear();
+
+                    var nextIndex = i + 1;
```
>var nextIndex = i + 1; [](http://example.com/codeflow?start=20&length=22)

I think the logic would be simpler to follow if we were to limit the purpose of this method to visiting IInterpolatedStringHandlerCreationOperation argument and returning it. The caller can check if it needs to break out of the loop.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2045,232 @@ private IOperation PeekOperand()
+                        var argumentPlaceholder = unwrapper(constructorArgument) switch
+                        {
+                            IInterpolatedStringHandlerArgumentPlaceholderOperation p => p,
+                            IConversionOperation { Operand: IInterpolatedStringHandlerArgumentPlaceholderOperation p } => p,
```
>Operand: IInterpolatedStringHandlerArgumentPlaceholderOperation p } => p, [](http://example.com/codeflow?start=51&length=73)

Similar comment as in hasPlaceholders.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -46,6 +46,8 @@ internal sealed partial class ControlFlowGraphBuilder : OperationVisitor<int?, I
         private IOperation? _forToLoopBinaryOperatorRightOperand;
         private IOperation? _currentAggregationGroup;
         private bool _forceImplicit; // Force all rewritten nodes to be marked as implicit regardless of their original state.
+        private PooledDictionary<IOperation, IOperation>? _placeholderDictionary;
```
>IOperation [](http://example.com/codeflow?start=33&length=10)

Is this always IInterpolatedStringHandlerArgumentPlaceholderOperation?

333fred replied:
This is gone now.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2045,232 @@ private IOperation PeekOperand()
+                }
+
+                var visitedArguments = ArrayBuilder<IOperation>.GetInstance(arguments.Length);
+                var placeholderKeys = ArrayBuilder<IOperation>.GetInstance();
```
>IOperation [](http://example.com/codeflow?start=51&length=10)

IInterpolatedStringHandlerArgumentPlaceholderOperation?

AlekseyTs replied:
>placeholderKeys [](http://example.com/codeflow?start=20&length=15)

placeholders?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2045,232 @@ private IOperation PeekOperand()
+                                    break;
+
+                                case InterpolatedStringArgumentPlaceholderKind.CallsiteArgument:
+                                    if (argumentPlaceholder.ArgumentIndex >= i)
```
>i [](http://example.com/codeflow?start=77&length=1)

visitedArguments.Count?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -4442,17 +4685,19 @@ InvocationOperation makeInvocationDroppingInstanceForStaticMethods(IMethodSymbol
                 if (arguments != null)
                 {
-                    return VisitArguments(arguments);
+                    (instanceOpt, var visitedArguments) = VisitArgumentsWithVisitedInstance(arguments, instanceOpt);
```
>arguments [](http://example.com/codeflow?start=92&length=9)

Can this ever contain an IInterpolatedStringHandlerCreationOperation?

333fred replied:
I don't think so, but it seemed more robust to always go through this codepath.

333fred replied:
I opted to assume there will never be an interpolated string handler in this code.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
 
         public override IOperation? VisitInterpolatedStringHandlerCreation(IInterpolatedStringHandlerCreationOperation operation, int? captureIdForResult)
```
>IInterpolatedStringHandlerCreationOperation [](http://example.com/codeflow?start=75&length=43)

Is it possible that this node is a child of an operation None and we fail to set up the placeholder replacements for it?

333fred replied:
I don't believe so: the only time we should have placeholders for the receiver/arguments should be an actual invocation of a method or property, and I don't believe we have any NoneOperations that would hit it today (at least, none that I see in the catch-all for `CSharpOperationFactory.Visit`).

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2045,232 @@ private IOperation PeekOperand()
+            return (visitedInstance == null ? null : PopOperand(), visitedArguments);
+        }
+
+        private void VisitAndPushArgumentsWithVisitedInstance(ImmutableArray<IArgumentOperation> arguments, IOperation? visitedInstance)
```
>VisitAndPushArgumentsWithVisitedInstance [](http://example.com/codeflow?start=21&length=40)

I think that, if we mostly rely on an infrastructure that we already have, the logic can be significantly simpler, and, perhaps, more robust. If there are `IInterpolatedStringHandlerCreationOperation`s among the arguments (and we don’t need to perform more detailed analysis about what placeholders are used during creation), we set up a current interpolation context for their rewrite. The context consists of the following information:
- The position in the `_evalStack` where the receiver and the arguments are going to start at
- Whether there is a receiver
- The set of IInterpolatedStringHandlerCreationOperation`s that are allowed to use the context.

Then we push the receiver on the stack, if any, and continue rewriting and pushing the arguments normally. Once all arguments are visited, we restore the previous interpolation context. 

When we reach an `IInterpolatedStringHandlerCreationOperation`, we can verify that the current interpolation context, if any, is set up to handle the node. And can set up an error recovery context, if we need to handle cases like that. We visit the creation node and, once we run into a `CallsiteArgument`/`CallsiteReceiver` `IInterpolatedStringHandlerArgumentPlaceholderOperation` node that has corresponding value on the stack (missing receivers and arguments out of order are not on the stack), we spill the eval stack and get the corresponding node from the `_evalStack` based on the context information.

We can discuss this approach in more details offline.

AlekseyTs replied:
```VisitInterpolatedStringHandlerCreation``` handles spilling of the stack so we don't need to do spilling for placeholders.

333fred replied:
I took this approach, and as a result all of the code in this chunk has been deleted. I'll resolve the comments in here as they no longer apply.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+
+                for (int i = arguments.Length - 1; i > 1; i--)
+                {
+                    if (arguments[i].ArgumentKind == ArgumentKind.Explicit)
```
>arguments[i].ArgumentKind == ArgumentKind.Explicit [](http://example.com/codeflow?start=24&length=50)

We are looking for ```is IInterpolatedStringHandlerArgumentPlaceholderOperation { PlaceholderKind: InterpolatedStringArgumentPlaceholderKind.TrailingValidityArgument }``` and that is what we should be using as the condition rather than trying to infer this information from properties that do not have primiry relation to the fact.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+            RegionBuilder resultRegion = CurrentRegionRequired;
+            var handlerCaptureId = captureIdForResult ?? GetNextCaptureId(resultRegion);
+            var previousHandlerId = _currentInterpolatedStringHandlerId;
+            _currentInterpolatedStringHandlerId = handlerCaptureId;
```
>_currentInterpolatedStringHandlerId [](http://example.com/codeflow?start=12&length=35)

This can be moved to the proposed interpolation context as a mutable state.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+            {
+                // Only successful constructor binds will have a trailing parameter
+                Debug.Assert(operation.HandlerCreation is IObjectCreationOperation);
+                outParameterFlowCapture = GetNextCaptureId(constructorRegion);
```
>outParameterFlowCapture [](http://example.com/codeflow?start=16&length=23)

This can be stored in the proposed interpolation context as a mutable state. Then we wouldn't need to set up the replacement in a side dictionary.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+
+            if (operation.HandlerCreationHasSuccessParameter)
+            {
+                RemovePlaceholder(outParameterPlaceholder!);
```
>RemovePlaceholder(outParameterPlaceholder!); [](http://example.com/codeflow?start=16&length=44)

It feels like this can be merged into the previous ```if```

333fred replied:
No longer present at all.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                ConditionalBranch(new FlowCaptureReferenceOperation(outParameterFlowCapture, outParameterPlaceholder.Syntax, outParameterPlaceholder.Type, constantValue: null), jumpIfTrue: false, resultBlock);
+                // else goto next block;
+                LeaveRegionsUpTo(resultRegion);
+                AppendNewBlock(new BasicBlockBuilder(BasicBlockKind.Block), linkToPrevious: true);
```
>AppendNewBlock(new BasicBlockBuilder(BasicBlockKind.Block), linkToPrevious: true); [](http://example.com/codeflow?start=16&length=82)

It looks like we simply do ```_currentBasicBlock = null;``` when we simply want to continue in the new linked block.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+            }
+            else
+            {
+                LeaveRegionsUpTo(resultRegion);
```
>LeaveRegionsUpTo(resultRegion); [](http://example.com/codeflow?start=16&length=31)

It looks like we can move this out of the ```if```.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                RemovePlaceholder(outParameterPlaceholder!);
+            }
+
+            var appendCalls = collectAppendCalls(operation);
```
>var [](http://example.com/codeflow?start=12&length=3)

The type of this local is not mentioned anywhere in this statement.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+            _currentInterpolatedStringHandlerId = previousHandlerId;
+            return new FlowCaptureReferenceOperation(handlerCaptureId, operation.Syntax, operation.Type, operation.GetConstantValue());
+
+            static ImmutableArray<IInterpolatedStringAppendOperation> collectAppendCalls(IInterpolatedStringHandlerCreationOperation creation)
```
>ImmutableArray<IInterpolatedStringAppendOperation> [](http://example.com/codeflow?start=19&length=50)

Why does this have to be an immutable array. The value is used locally and it looks like leaving it as a builder would be just fine.

333fred replied:
Personal preference of not having to then manage the lifetime of the builder across functions.

AlekseyTs replied:
> Personal preference of not having to then manage the lifetime of the builder across functions.

I do not think the gained simplicity outweighs the fact that we unnecessary allocate an array. Creation of the builder can be pulled to the call site to make lifetime management contract clear.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                static void pushLeftNodes(IInterpolatedStringAdditionOperation addition, ArrayBuilder<IInterpolatedStringAdditionOperation> stack)
+                {
+                    IInterpolatedStringAdditionOperation? current = addition;
+                    while (current != null)
```
>while (current != null) [](http://example.com/codeflow?start=20&length=23)

We know that we will do at least one iteration. In cases like that a post condition should be used, i.e. ```do while```

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                var stack = ArrayBuilder<IInterpolatedStringAdditionOperation>.GetInstance();
+                pushLeftNodes((IInterpolatedStringAdditionOperation)creation.Content, stack);
+
+                while (stack.TryPop(out var currentAddition))
```
>var [](http://example.com/codeflow?start=40&length=3)

Please spell out the type

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                        // No matter the result, we're going to the result block next. So just visit the statement, and if the current block can be
+                        // combined with the result block, the compaction machinery will take care of it
+                        VisitStatement(appendCall.AppendCall);
+                        AppendNewBlock(resultBlock, linkToPrevious: true);
```
>AppendNewBlock(resultBlock, linkToPrevious: true); [](http://example.com/codeflow?start=24&length=50)

```_currentBasicBlock = null;```

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                    {
+                        // if (!appendCall()) goto result else goto next block
+                        ConditionalBranch(VisitRequired(appendCall.AppendCall), jumpIfTrue: false, resultBlock);
+                        AppendNewBlock(new BasicBlockBuilder(BasicBlockKind.Block), linkToPrevious: true);
```
>AppendNewBlock(new BasicBlockBuilder(BasicBlockKind.Block), linkToPrevious: true); [](http://example.com/codeflow?start=24&length=82)

```_currentBasicBlock = null;```

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                    if (i == appendCallsLength - 1 && operation.HandlerCreationHasSuccessParameter)
+                    {
+                        Debug.Assert(resultBlock != null);
+                        AppendNewBlock(resultBlock, linkToPrevious: true);
```
>AppendNewBlock(resultBlock, linkToPrevious: true) [](http://example.com/codeflow?start=24&length=49)

```_currentBasicBlock = null;```

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                LeaveRegion();
+            }
+
+            if (appendCallsLength == 0 && operation.HandlerCreationHasSuccessParameter)
```
>appendCallsLength == 0 && operation.HandlerCreationHasSuccessParameter [](http://example.com/codeflow?start=16&length=70)

Manual management of basic blocks here feels unnecessary complicated. I think all we really need to do is to ```_currentBasicBlock = null;``` after conditional branches and here unconditionally append ```resultBlock``` if we have one.

AlekseyTs replied:
This doesn't look addressed. The suggestion isn't followed, there is stil manual management that is not related to branching.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6561,190 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                    else
+                    {
+                        // if (!appendCall()) goto result else goto next block
+                        ConditionalBranch(VisitRequired(appendCall.AppendCall), jumpIfTrue: false, resultBlock);
```
>VisitRequired(appendCall.AppendCall) [](http://example.com/codeflow?start=42&length=36)

Consider taking this out of the ```if``` and explisitly calling ```AddStatement``` for the unconditional case.

AlekseyTs replied:
It looks like ```VisitRequired(appendCall.AppendCall);``` wasn't pulled far enough. We can go one more level. I think there is a benefit in visiting the call one way regardless of how the result is going to be used.

AlekseyTs commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -814,6 +815,11 @@ PooledHashSet<CaptureId> adjustAndGetEntryState(ArrayBuilder<PooledHashSet<Captu
 
+                    if (reference.IsInitialization)
+                    {
+                        state.Add(id);
```
>state.Add(id); [](http://example.com/codeflow?start=24&length=14)

I think we should assert that ```Add``` returns true similar to how we do that on line 413.

AlekseyTs commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -814,6 +815,11 @@ PooledHashSet<CaptureId> adjustAndGetEntryState(ArrayBuilder<PooledHashSet<Captu
 
+                    if (reference.IsInitialization)
+                    {
+                        state.Add(id);
```
>state.Add(id); [](http://example.com/codeflow?start=24&length=14)

Is there a need to perfom checks below? It looks like we don't do that for IFlowCaptureOperation

333fred replied:
I don't think so.

AlekseyTs commented:
>                     referencedIds.Add(id);

The initialization isn't a reference, I think. So we probably don't want to record it like that. Same for isLongLivedCaptureReference below.

---
Refers to: src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:811 in 3c58e1c. [](commit_id = 3c58e1c1f396b3669408d7dc008b37896c5486ea, deletion_comment = False)

AlekseyTs commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -892,6 +901,54 @@ bool isObjectInitializerInitializedObjectTarget(IFlowCaptureReferenceOperation r
+                                   Parent: IFlowCaptureOperation,
+                                   Constructor.ContainingType: INamedTypeSymbol ctorContainingType,
+                                   Arguments: { Length: >= 3 } arguments,
+                                   Syntax: CSharp.Syntax.InterpolatedStringExpressionSyntax or CSharp.Syntax.BinaryExpressionSyntax
```
>CSharp.Syntax.BinaryExpressionSyntax [](http://example.com/codeflow?start=95&length=36)

Should we make this stronger and actuall check that there is an interpolation addition?

333fred replied:
We could, but I don't think it's worth it. I don't know of any other way in C# to turn a binary expression syntax into an `IObjectCreationOperation` node, and the extra complexity from walking the tree (through any parenthesized/suppressed nodes) isn't worth the small extra verification.

AlekseyTs commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -892,6 +901,54 @@ bool isObjectInitializerInitializedObjectTarget(IFlowCaptureReferenceOperation r
+                                   Parent: IFlowCaptureOperation,
+                                   Constructor.ContainingType: INamedTypeSymbol ctorContainingType,
+                                   Arguments: { Length: >= 3 } arguments,
+                                   Syntax: CSharp.Syntax.InterpolatedStringExpressionSyntax or CSharp.Syntax.BinaryExpressionSyntax
```
>CSharp.Syntax.InterpolatedStringExpressionSyntax [](http://example.com/codeflow?start=43&length=48)

Should we worry about parenthesis and supressions?

333fred replied:
Probably.

AlekseyTs commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -892,6 +901,54 @@ bool isObjectInitializerInitializedObjectTarget(IFlowCaptureReferenceOperation r
+                       {
+                           Parent: IArgumentOperation
+                           {
+                               Parent: IObjectCreationOperation
```
> Parent: IObjectCreationOperation [](http://example.com/codeflow?start=30&length=33)

Can there be a conversion on top?

AlekseyTs commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -892,6 +901,54 @@ bool isObjectInitializerInitializedObjectTarget(IFlowCaptureReferenceOperation r
+#pragma warning disable IDE0055 // Fix formatting
+                return reference is
+                       {
+                           Parent: IArgumentOperation
```
>Parent: IArgumentOperation [](http://example.com/codeflow?start=27&length=26)

Can there be a conversion on top?

AlekseyTs commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -824,8 +830,11 @@ PooledHashSet<CaptureId> adjustAndGetEntryState(ArrayBuilder<PooledHashSet<Captu
+                                     isObjectInitializerInitializedObjectTarget(reference) ||
+                                     isInterpolatedStringArgumentCapture(reference) ||
+                                     isInterpolatedStringHandlerCapture(reference)) &&
+                                 (block.EnclosingRegion.EnclosingRegion.CaptureIds.Contains(id) ||
```
>block.EnclosingRegion.EnclosingRegion.CaptureIds.Contains(id) [](http://example.com/codeflow?start=34&length=61)

Would it be possible to bundle this with ```isObjectInitializerInitializedObjectTarget(reference)``` (as it used to be). And bundle ```(block.EnclosingRegion.EnclosingRegion?.EnclosingRegion.CaptureIds.Contains(id) ?? false)``` with interpolation captures. The goal is to make the check stronger.

333fred replied:
Yes, but I'm going to break this condition up to do so. It's at my complexity limit now, and trying to nest more things breaks my brain reading it.

333fred replied:
Actually, I'm not sure what you mean by the first part. If I move the `block.EnclosingRegion.EnclosingRegion.CaptureIds.Contains(id)` check, it weakens the rest of the checks here by allowing them to exist anywhere in the graph.

333fred replied:
I did my best interpretation of what I think you meant. Let me know if it's not what you had in mind.

AlekseyTs replied:
> I did my best interpretation of what I think you meant. Let me know if it's not what you had in mind.

Sorry for the confusion. I think I slightly misread the original condition, but you adapted the suggestion correctly.

AlekseyTs commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -824,8 +830,11 @@ PooledHashSet<CaptureId> adjustAndGetEntryState(ArrayBuilder<PooledHashSet<Captu
-                                     isObjectInitializerInitializedObjectTarget(reference)) &&
-                                 block.EnclosingRegion.EnclosingRegion.CaptureIds.Contains(id)),
+                                     isObjectInitializerInitializedObjectTarget(reference) ||
+                                     isInterpolatedStringArgumentCapture(reference) ||
```
>isInterpolatedStringArgumentCapture [](http://example.com/codeflow?start=37&length=35)

It looks like the reference for the special out argument shouldn't be the subject for the relaxed check. It feels like the reference should belong to the enclosing region.

AlekseyTs commented:
>                 Leaving: {R2}

Is it worth optimizing this branch out? We could use IDiscardOperation in the argument list above.

---
Refers to: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:4349 in 3c58e1c. [](commit_id = 3c58e1c1f396b3669408d7dc008b37896c5486ea, deletion_comment = False)

AlekseyTs commented:
>         public void NoStringContent()

This test is never going to fail.

---
Refers to: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:5817 in 3c58e1c. [](commit_id = 3c58e1c1f396b3669408d7dc008b37896c5486ea, deletion_comment = False)

AlekseyTs commented:
Done with review pass (commit 7)

333fred commented:
>         public void NoStringContent()

I folded that into _16, but forgot to delete the stub.

---
In reply to: [961330742](https://github.com/dotnet/roslyn/pull/57483#issuecomment-961330742) [](http://example.com/codeflow?ancestors=961330742,961330742)

---
Refers to: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:5817 in 3c58e1c. [](commit_id = 3c58e1c1f396b3669408d7dc008b37896c5486ea, deletion_comment = False)

333fred commented:
>                 Leaving: {R2}

I don't think so. It would require extra message passing to the placeholder visit with the new architecture, for a scenario that is pretty uncommon (very common for my tests, but unlikely to be a common case in the wild).

---
In reply to: [961328809](https://github.com/dotnet/roslyn/pull/57483#issuecomment-961328809) [](http://example.com/codeflow?ancestors=961328809,961328809)

---
Refers to: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:4349 in 3c58e1c. [](commit_id = 3c58e1c1f396b3669408d7dc008b37896c5486ea, deletion_comment = False)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [961332437](https://github.com/dotnet/roslyn/pull/57483#issuecomment-961332437) [](http://example.com/codeflow?ancestors=961332437,961332437)

AlekseyTs commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -807,6 +808,14 @@ PooledHashSet<CaptureId> adjustAndGetEntryState(ArrayBuilder<PooledHashSet<Captu
+                    {
+                        AssertTrueWithGraph(state.Add(id), $"Multiple initialization of [{id}]", finalGraph);
+                        AssertTrueWithGraph(block.EnclosingRegion.CaptureIds.Contains(id), $"Flow capture initialization [{id}] should come from the containing region.", finalGraph);
+                        return;
```
> return; [](http://example.com/codeflow?start=23&length=8)

It feels like continue is more appropriate.

AlekseyTs commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -819,14 +828,34 @@ PooledHashSet<CaptureId> adjustAndGetEntryState(ArrayBuilder<PooledHashSet<Captu
-                        $"Operation [{operationIndex}] in [{getBlockId(block)}] uses capture [{id.Value}] from another region. Should the regions be merged?", finalGraph);
+                    if (block.EnclosingRegion.CaptureIds.Contains(id) || longLivedIds.Contains(id))
+                    {
+                        return;
```
>return; [](http://example.com/codeflow?start=24&length=7)

```continue```?

AlekseyTs commented on src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -892,6 +921,58 @@ bool isObjectInitializerInitializedObjectTarget(IFlowCaptureReferenceOperation r
+                do
+                {
+                    containingArgument = containingArgument.Parent;
+                } while (containingArgument is not (null or IArgumentOperation));
```
>} while (containingArgument is not (null or IArgumentOperation)); [](http://example.com/codeflow?start=16&length=65)

This looks like a violation of a style. Shouldn't braces be on their own lines?

333fred replied:
I use `do/while` loops so infrequently I haven't broken my Java habits with them.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6358,182 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+
+            // Any placeholders for arguments should have already been created, except for the out parameter if it exists.
+            int outParameterFlowCapture = -1;
+            IOperation? outParameterPlaceholder = null;
```
>IOperation [](http://example.com/codeflow?start=12&length=10)

IInterpolatedStringHandlerArgumentPlaceholderOperation

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6358,182 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                        // No matter the result, we're going to the result block next. So just visit the statement, and if the current block can be
+                        // combined with the result block, the compaction machinery will take care of it
+                        AddStatement(visitedAppendCall);
+                        _currentBasicBlock = null;
```
>_currentBasicBlock = null; [](http://example.com/codeflow?start=24&length=26)

What is the point of doing this?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6358,182 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                    if (i == appendCallsLength - 1 && operation.HandlerCreationHasSuccessParameter)
+                    {
+                        Debug.Assert(resultBlock != null);
+                        _currentBasicBlock = null;
```
>_currentBasicBlock = null; [](http://example.com/codeflow?start=24&length=26)

What is the point of doing this?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6358,182 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                    }
+                }
+                LeaveRegionsUpTo(appendRegion);
+                LeaveRegion();
```
It looks like these two statements can be replaced with ```LeaveRegionsUpTo(resultRegion)```, which actually makes it clear where we are going to end up.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.InterpolatedStringContext.cs:
```diff
@@ -0,0 +1,108 @@
+            }
+        }
+
+        private static void AssertContainingContextIsForThisCreation(IOperation placeholderOperation, InterpolatedStringHandlerContext context)
```
>InterpolatedStringHandlerContext context [](http://example.com/codeflow?start=102&length=40)

It looks like we allways pass ```_currentInterpolatedStringHandlerContext``` for this parameter. What is the benefit in having the parameter vs. making this an instance method.

AlekseyTs replied:
>AssertContainingContextIsForThisCreation [](http://example.com/codeflow?start=28&length=40)

```Conditional("DEBUG")```?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6337,7 +6543,88 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+            IOperation? tryGetArgumentOrReceiver(int argumentIndex)
+            {
+                int startingStackDepth = _currentInterpolatedStringHandlerContext.StartingStackDepth;
+                if (startingStackDepth >= _evalStack.Count)
```
>if (startingStackDepth >= _evalStack.Count) [](http://example.com/codeflow?start=16&length=43)

It looks like we are not handling the ```-1``` case, which is apparently can happen in an error recovery scenario.

333fred replied:
I don't believe this case is currently possible.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6337,7 +6543,88 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                {
+                    Debug.Assert(_currentInterpolatedStringHandlerContext.HasReceiver);
+
+                    // The receiver is always directly at the starting stack depth: there is no additional frame pushed on at
```
>// The receiver is always directly at the starting stack depth: there is no additional frame pushed on at [](http://example.com/codeflow?start=20&length=105)

I don't think we have to make this assumption. In fact, the implementation would be more robust if we wouldn't. The solution looks simple, when ```_currentInterpolatedStringHandlerContext.HasReceiver```, increment ```argumentIndex``` and proceed to the loop below.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6337,7 +6543,88 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                // We need to walk forward until we hit a stack frame. If we hit a stack frame, then we're out of the allowable context
+                // for the current method (meaning that the argument is missing), and we've hit the handler constructor arguments themselves.
+                int operationsSeen = 0;
+                for (int currentStackDepth = startingStackDepth + (_currentInterpolatedStringHandlerContext.HasReceiver ? 1 : 0);
```
>+ (_currentInterpolatedStringHandlerContext.HasReceiver ? 1 : 0 [](http://example.com/codeflow?start=64&length=63)

We wouldn't need this adjustment with the suggestion above.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6337,7 +6543,88 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+                    return operation;
+                }
+
+                // We need to walk forward until we hit a stack frame. If we hit a stack frame, then we're out of the allowable context
```
>If we hit a stack frame, then we're out of the allowable context [](http://example.com/codeflow?start=71&length=64)

I think tt wold be more robust to not rely on this assumption. In order to achieve that we could simply record the stack depth at the time we enter ```VisitInterpolatedStringHandlerCreation``` in the context. That is the limit we don't want to go over.

333fred replied:
I don't think that would be more robust: in fact, I think it's more robust to ensure we stop on any new stack frame. A new stack frame means that we're in a new function call, specifically the the stack frame for the current `InterpolatedStringHandlerCreationOperation` being processed.

AlekseyTs replied:
The robust way is to not infer where to stop, but to know where to stop. If you think getting the stack depth in ```VisitInterpolatedStringHandlerCreation``` is still not strong enough, we can record position of each ```IInterpolatedStringHandlerCreationOperation``` in the argument list in the context and base logic on that. If you want to assert that there are no frames, that would be fine, but I don't think we should base our algorithm on that.

AlekseyTs commented:
Do we understand why CI jobs are failing?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.InterpolatedStringContext.cs:
```diff
@@ -0,0 +1,108 @@
+{
+    internal partial class ControlFlowGraphBuilder
+    {
+        private readonly struct InterpolatedStringHandlerContext
```
>InterpolatedStringHandlerContext [](http://example.com/codeflow?start=32&length=32)

It feels like we have a certain amount of over-engineering here. In particular:
- I don’t think we would want to inline this structure into ```ControlFlowGraphBuilder```. A class would work just fine, especially given that ```IInterpolatedStringHandlerCreationOperation```s won’t be common.
- The implementation needs to worry about default values and has to explicitly track the state to recognize them.
- Flip-flopping between “single” and “multiple” mode adds a fair amount of complexity to the implementation and makes it harder to keep track of interaction between different parts of the builder. I haven’t finished reviewing all changes, but it looks like the only advantage we get from the “single” mode is the ability to assert the ``` IInterpolatedStringHandlerCreationOperation``` node being visited. So, it is not instrumental to the algorithm overall. In my opinion the added complexity does not outweigh the benefit to be able to do ```Debug.Assert```. The same can be achieved by adding a DEBUG-only field to the context, that is set/cleared in ```VisitInterpolatedStringHandlerCreation```.
- Flip-flopping between the ```IInterpolatedStringHandlerCreationOperation``` and ```IArgumentOperation``` also adds complexity.

Bottom line, I think we should switch to class to represent the context. Some portion of the, context that is initialized when we preparing to visit arguments, is read-only. The other part of the context is mutable and specific to particular ```IInterpolatedStringHandlerCreationOperation``` node we are visiting. That portion is set/cleared in ```VisitInterpolatedStringHandlerCreation```. The parts can be split in two different classes and can be stored in two different fields of ```ControlFlowGraphBuilder```, but I do not see any big advantage of doing this. All context that ```ControlFlowGraphBuilder``` is mutable. I don’t see much difference between modifying fields directly stored in ```ControlFlowGraphBuilder``` and modifying fields stored in classes stored in fields of ```ControlFlowGraphBuilder```. So, going after read-only fields is a non-goal when we are talking about the context.

333fred replied:
I can try to simplify the implementation by splitting it up into two classes, but I think a partially-mutable class where some of the fields apply to a specific set of interpolated string creations and some of the fields apply to the _current_ creation (regardless of whether the current creation is in that previous list) is more confusing than the current implementation.

jaredpar commented:
FYI: the correctness failure looks real here 

```
ID=csc 7072 TID=5: Debug.Assert failed with message: Fail:
System.NullReferenceException: Object reference not set to an instance of an object.
at System.Collections.Immutable.ImmutableArray`1.GetEnumerator()
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.VisitAndPushArray[T](ImmutableArray`1 array, Func`2 unwrapper) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 1982
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.VisitArray[T](ImmutableArray`1 originalArray, Func`2 unwrapper, Func`4 wrapper) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 2016
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.<AddDisposingFinally>g__tryDispose|136_0(IOperation value, <>c__DisplayClass136_0& ) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 4025
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.AddDisposingFinally(IOperation resource, Boolean requiresRuntimeConversion, ITypeSymbol iDisposable, IMethodSymbol disposeMethod, ImmutableArray`1 disposeArguments, Boolean isAsynchronous) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 4004
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.<HandleUsingOperationParts>g__processResource|135_2(IOperation resource, ArrayBuilder`1 resourceQueueOpt, <>c__DisplayClass135_0& ) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 3955
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.<HandleUsingOperationParts>g__processQueue|135_0(ArrayBuilder`1 resourceQueueOpt, <>c__DisplayClass135_0& ) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 3877
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.HandleUsingOperationParts(IOperation resources, IOperation body, IMethodSymbol disposeMethod, ImmutableArray`1 disposeArguments, ImmutableArray`1 locals, Boolean isAsynchronous) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 3839
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.VisitUsingVariableDeclarationOperation(IUsingDeclarationOperation operation, ReadOnlySpan`1 statements) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 7382
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.VisitStatementsOneOrAll(IOperation operation, ImmutableArray`1 statements, Int32 startIndex) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 1440
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.VisitStatements(ImmutableArray`1 statements) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 1416
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.VisitBlock(IBlockOperation operation, Nullable`1 captureIdForResult) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 1384
at Microsoft.CodeAnalysis.Operations.BlockOperation.Accept[TArgument,TResult](OperationVisitor`2 visitor, TArgument argument) in /_/src/Compilers/Core/Portable/Generated/Operations.Generated.cs:line 3472
at Microsoft.CodeAnalysis.Operations.OperationVisitor`2.Visit(IOperation operation, TArgument argument) in /_/src/Compilers/Core/Portable/Generated/Operations.Generated.cs:line 8573
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.VisitStatement(IOperation operation) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 1244
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.VisitMethodBodies(IBlockOperation blockBody, IBlockOperation expressionBody) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 1512
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.VisitMethodBodyBaseOperation(IMethodBodyBaseOperation operation) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 1505
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.VisitMethodBodyOperation(IMethodBodyOperation operation, Nullable`1 captureIdForResult) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 1498
at Microsoft.CodeAnalysis.Operations.MethodBodyOperation.Accept[TArgument,TResult](OperationVisitor`2 visitor, TArgument argument) in /_/src/Compilers/Core/Portable/Generated/Operations.Generated.cs:line 6753
at Microsoft.CodeAnalysis.Operations.OperationVisitor`2.Visit(IOperation operation, TArgument argument) in /_/src/Compilers/Core/Portable/Generated/Operations.Generated.cs:line 8573
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.VisitStatement(IOperation operation) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 1244
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraphBuilder.Create(IOperation body, ControlFlowGraph parent, ControlFlowRegion enclosing, CaptureIdDispenser captureIdDispenser, Context& context) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:line 132
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraph.CreateCore(IOperation operation, String argumentNameForException, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/Operations/ControlFlowGraph.cs:line 182
Stack Trace
at Microsoft.CodeAnalysis.CommandLine.ExitingTraceListener.Exit(String originalMessage)
at Microsoft.CodeAnalysis.CommandLine.ExitingTraceListener.WriteLine(String message)
at System.Diagnostics.TraceInternal.Fail(String message)
at System.Diagnostics.Debug.Assert(Boolean condition, String message)
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraph.CreateCore(IOperation operation, String argumentNameForException, CancellationToken cancellationToken)
at Microsoft.CodeAnalysis.FlowAnalysis.ControlFlowGraph.Create(IMethodBodyOperation methodBody, CancellationToken cancellationToken)
at Analyzer.Utilities.Extensions.IOperationExtensions.CreateControlFlowGraph(IOperation operation)
at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
at Analyzer.Utilities.Extensions.IOperationExtensions.TryGetEnclosingControlFlowGraph(IOperation operation, ControlFlowGraph& cfg)
at Analyzer.Utilities.Extensions.OperationBlocksExtensions.GetControlFlowGraph(ImmutableArray`1 operationBlocks)
at Microsoft.NetCore.Analyzers.InteropServices.PlatformCompatibilityAnalyzer.<>c__DisplayClass35_0.<AnalyzeOperationBlock>b__1(OperationBlockAnalysisContext context)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.<>c__67`2.<ExecuteBlockActions>b__67_1(ValueTuple`2 data)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows_NoLock[TArg](DiagnosticAnalyzer analyzer, Action`1 analyze, TArg argument, Nullable`1 info)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteAndCatchIfThrows[TArg](DiagnosticAnalyzer analyzer, Action`1 analyze, TArg argument, Nullable`1 info)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteBlockActions[TBlockAction,TNodeStateData](PooledHashSet`1 blockActions, SyntaxNode declaredNode, ISymbol declaredSymbol, DiagnosticAnalyzer analyzer, SemanticModel semanticModel, ImmutableArray`1 operationBlocks, Action`1 addDiagnostic, Func`2 isSupportedDiagnostic, BlockAnalyzerStateData`2 analyzerState)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.ExecuteBlockActionsCore[TBlockStartAction,TBlockAction,TNodeAction,TNodeStateData,TNode,TLanguageKindEnum](IEnumerable`1 startActions, IEnumerable`1 actions, IEnumerable`1 endActions, DiagnosticAnalyzer analyzer, SyntaxNode declaredNode, ISymbol declaredSymbol, ImmutableArray`1 executableBlocks, Func`2 getNodesToAnalyze, SemanticModel semanticModel, Func`2 getKind, BlockAnalyzerStateData`2 analyzerState, Boolean isGeneratedCode)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerExecutor.TryExecuteOperationBlockActions(IEnumerable`1 operationBlockStartActions, IEnumerable`1 operationBlockActions, IEnumerable`1 operationBlockEndActions, DiagnosticAnalyzer analyzer, SyntaxNode declaredNode, ISymbol declaredSymbol, ImmutableArray`1 operationBlocks, ImmutableArray`1 operations, SemanticModel semanticModel, Int32 declarationIndex, AnalysisScope analysisScope, AnalysisState analysisState, Boolean isGeneratedCode)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.<>c__DisplayClass12_0.<TryExecuteDeclaringReferenceActions>g__executeOperationsBlockActions|7(ImmutableArray`1 operationBlocksToAnalyze, ImmutableArray`1 operationsToAnalyze, IEnumerable`1 codeBlockActions)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.<>c__DisplayClass12_0.<TryExecuteDeclaringReferenceActions>g__executeExecutableCodeActions|3()
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.TryExecuteDeclaringReferenceActions(SyntaxReference decl, Int32 declarationIndex, SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, AnalysisState analysisState, GroupedAnalyzerActions coreActions, GroupedAnalyzerActions additionalPerSymbolActions, Boolean shouldExecuteSyntaxNodeActions, Boolean shouldExecuteOperationActions, Boolean shouldExecuteCodeBlockActions, Boolean shouldExecuteOperationBlockActions, Boolean isInGeneratedCode, CancellationToken cancellationToken)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.TryExecuteDeclaringReferenceActions(SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, AnalysisState analysisState, Boolean isGeneratedCodeSymbol, IGroupedAnalyzerActions additionalPerSymbolActions, CancellationToken cancellationToken)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.<TryProcessSymbolDeclaredAsync>d__132.MoveNext()
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.TryProcessSymbolDeclaredAsync(SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, AnalysisState analysisState, CancellationToken cancellationToken)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.<TryProcessEventCoreAsync>d__131.MoveNext()
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.Start[TStateMachine](TStateMachine& stateMachine)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.TryProcessEventCoreAsync(CompilationEvent compilationEvent, AnalysisScope analysisScope, AnalysisState analysisState, CancellationToken cancellationToken)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.<ProcessEventAsync>d__129.MoveNext()
at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ProcessEventAsync(CompilationEvent e, AnalysisScope analysisScope, AnalysisState analysisState, CancellationToken cancellationToken)
at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.<ProcessCompilationEventsCoreAsync>d__128.MoveNext()
at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2023,40 @@ private IOperation PeekOperand()
+            if (arguments.WhereAsArray(arg => arg.Value is IInterpolatedStringHandlerCreationOperation) is { IsDefaultOrEmpty: false } interpolatedStrings)
+            {
+                _currentInterpolatedStringHandlerContext = new InterpolatedStringHandlerContext(
+                    interpolatedStrings,
```
>interpolatedStrings [](http://example.com/codeflow?start=20&length=19)

I do not see a benefit in recording ```IArgumentOperation```s here instead of ```IInterpolatedStringHandlerCreationOperation```s

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2022,16 +2023,40 @@ private IOperation PeekOperand()
             return argument.Value;
         }
 
+        private static readonly Func<IOperation, IOperation> Identity = IdentityDoNotCaptureDirectly;
```
>Identity [](http://example.com/codeflow?start=61&length=8)

It doesn't look like this is used anywhere

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -3997,7 +4022,7 @@ private void AddDisposingFinally(IOperation resource, bool requiresRuntimeConver
                 if (method != null)
                 {
-                    var args = disposeMethod is object ? VisitArguments(disposeArguments) : ImmutableArray<IArgumentOperation>.Empty;
+                    var args = VisitArray(disposeArguments, UnwrapArgument, RewriteArgumentFromArray);
```
> var args = VisitArray(disposeArguments, UnwrapArgument, RewriteArgumentFromArray); [](http://example.com/codeflow?start=19&length=83)

It is not obvious why we need to make changes here, especially removal of the conditional logic looks suspicious

AlekseyTs replied:
I think that in the long term it would be better to keep calling VisitArguments with appropriate value for the second parameter.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -4442,17 +4469,18 @@ InvocationOperation makeInvocationDroppingInstanceForStaticMethods(IMethodSymbol
                 if (arguments != null)
                 {
-                    return VisitArguments(arguments);
+                    return VisitArray(arguments, UnwrapArgument, RewriteArgumentFromArray);
```
>VisitArray [](http://example.com/codeflow?start=27&length=10)

I think that in the long term it would be better to keep calling VisitArguments with appropriate value for the second parameter.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -5682,7 +5711,7 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
         {
             EvalStackFrame frame = PushStackFrame();
             EvalStackFrame argumentsFrame = PushStackFrame();
-            ImmutableArray<IArgumentOperation> visitedArgs = VisitArguments(operation.Arguments);
```
>VisitArguments [](http://example.com/codeflow?start=61&length=14)

It is not clear why we simply didn't pass ```false``` as the second argument.

333fred replied:
Because the first approach needed this version of the call, and I didn't fully revert this bit when I took the new approach.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6322,12 +6358,182 @@ private IOperation VisitNoneOperationExpression(IOperation operation)
+            int outParameterFlowCapture = -1;
+            IOperation? outParameterPlaceholder = null;
+
+            if (operation.HandlerCreationHasSuccessParameter)
```
>if (operation.HandlerCreationHasSuccessParameter) [](http://example.com/codeflow?start=12&length=49)

We can call ```collectAppendCalls``` before this ```if``` and, if the set is empty, not enter the ```if```, leaving ```outParameterFlowCapture``` == -1. Then we suppress creation of conditional branch in the ```if``` below when ```outParameterFlowCapture == -1```. Then we can adjust ```VisitInterpolatedStringHandlerArgumentPlaceholder``` to create ```IDiscardOperation``` in this case. This way we can eliminate an unnecessary branch, see ```InterpolatedStringHandlerConversionFlow_16``` for example. This work doesn't look prohibiteviley complex or expensive to me.

333fred replied:
I filed [https://github.com/dotnet/roslyn/issues/57611](https://github.com/dotnet/roslyn/issues/57611) to track doing this work at some point in the future.

AlekseyTs commented:
Done with review pass (commit 8)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [962158460](https://github.com/dotnet/roslyn/pull/57483#issuecomment-962158460) [](http://example.com/codeflow?ancestors=962158460,962158460)

AlekseyTs commented:
It looks like the build is broken

AlekseyTs commented:
>                 ImmutableArray<IArgumentOperation> visitedArguments = VisitArguments(arguments);

We probably want to pass ```false``` as the second parameter here.

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:5721 in 3ecd8ad. [](commit_id = 3ecd8ad96a1ae25fb3b2550ce5a1610091924166, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -6343,12 +6398,191 @@
         public override IOperation? VisitInterpolatedStringHandlerCreation(IInterpolatedStringHandlerCreationOperation operation, int? captureIdForResult)
         {
-            return VisitNoneOperation(operation, captureIdForResult);
+            // We turn the interpolated string into a call to create the handler type, a series of append calls (potentially with branches, depending on the
```
>Contains [](http://example.com/codeflow?start=95&length=8)

It looks like we are leaving the room for the ```Contains``` to not find the match. I.e. we will not assert here, at the same time, it looks like we will use _currentInterpolatedStringHandlerArgumentContext regardless.

AlekseyTs commented:
Done with review pass (commit 10)

AlekseyTs commented on src/Compilers/Core/Portable/Collections/SpanExtensions.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    internal static class SpanExtensions
+    {
+        public static bool Any<T>(this ReadOnlySpan<T> span, Func<T, bool> predicate)
```
>Any [](http://example.com/codeflow?start=27&length=3)

Do we still use this helper?

AlekseyTs reviewed (approved):
LGTM (commit 12), assuming CI is passing.

333fred commented:
/azp run
